File: ios/App/App/IsolatedModules/FileExplorer.swift
//
//  FileExplorer.swift
//  App
//
//  Created by Julia Samol on 08.02.23.
//

import Foundation

// MARK: FileExplorer

struct FileExplorer {
    static let shared: FileExplorer = .init()
    
    private let assetsExplorer: AssetsExplorer
    private let documentExplorer: DocumentExplorer
    
    private let fileManager: FileManager
    
    init(fileManager: FileManager = .default) {
        self.assetsExplorer = .init(fileManager: fileManager)
        self.documentExplorer = .init(fileManager: fileManager)
        self.fileManager = fileManager
    }
    
    func readIsolatedModulesScript() throws -> Data {
        try assetsExplorer.readIsolatedModulesScript()
    }
    
    func loadAssetModules() throws -> [JSModule] {
        try loadModules(using: assetsExplorer, creatingModuleWith: JSModule.Asset.init).map { .asset($0) }
    }
    
    func loadInstalledModule(_ identifier: String) throws -> JSModule {
        let manifest = try documentExplorer.readModuleManifest(identifier)
        
        return .installed(try loadModule(identifier, fromManifest: manifest, creatingModuleWith: JSModule.Instsalled.init))
    }
    
    func loadInstalledModules() throws -> [JSModule] {
        try loadModules(using: documentExplorer, creatingModuleWith: JSModule.Instsalled.init).map { .installed($0) }
    }
    
    func loadPreviewModule(atPath path: String, locatedIn directory: Directory) throws -> JSModule {        
        guard let directory = fileManager.getDirectory(from: directory),
              let url = fileManager.urls(for: directory, in: .userDomainMask).first?.appendingPathComponent(path) else {
            throw Error.invalidDirectory
        }
        
        let identifier = url.pathComponents.last ?? "module"
        let manifest = try fileManager.contents(at: url.appendingPathComponent(FileExplorer.manifestFilename))
        
        return .preview(try loadModule(identifier, fromManifest: manifest) { identifier, namespace, preferredEnvironment, sources in
            JSModule.Preview(identifier: identifier, namespace: namespace, preferredEnvironment: preferredEnvironment, sources: sources, path: url)
        })
    }
    
    func readModuleSources(_ module: JSModule) throws -> [Data] {
        switch module {
        case .asset(let asset):
            return try assetsExplorer.readModuleSources(asset)
        case .installed(let installed):
            return try documentExplorer.readModuleSources(installed)
        case .preview(let preview):
            return try preview.sources.lazy.map { try fileManager.contents(at: preview.path.appendingPathComponent($0)) }
        }
    }
    
    func readModuleManifest(_ module: JSModule) throws -> Data {
        switch module {
        case .asset(let asset):
            return try assetsExplorer.readModuleManifest(asset.identifier)
        case .installed(let installed):
            return try documentExplorer.readModuleManifest(installed.identifier)
        case .preview(let preview):
            return try fileManager.contents(at: preview.path.appendingPathComponent(FileExplorer.manifestFilename))
        }
    }
    
    private func loadModules<T: JSModuleProtocol, E: DynamicSourcesExplorer>(
        using explorer: E,
        creatingModuleWith moduleInit: (_ identifier: String, _ namespace: String?, _ preferredEnvironment: JSEnvironmentKind, _ sources: [String]) -> T
    ) throws -> [T] where E.T == T {
        try explorer.listModules().map { module in
            try loadModule(module, fromManifest: try explorer.readModuleManifest(module), creatingModuleWith: moduleInit)
        }
    }
    
    private func loadModule<T: JSModuleProtocol>(
        _ identifier: String,
        fromManifest manifestData: Data,
        creatingModuleWith moduleInit: (_ identifier: String, _ namespace: String?, _ preferredEnvironment: JSEnvironmentKind, _ sources: [String]) -> T
    ) throws -> T {
        let jsonDecoder = JSONDecoder()
    
        let manifest = try jsonDecoder.decode(ModuleManifest.self, from: manifestData)
        let namespace = manifest.src?.namespace
        let preferredEnvironment = manifest.jsenv?.ios ?? .webview
        let sources: [String] = manifest.include.compactMap { source in
            guard source.hasSuffix(".js") else { return nil }
            return source
        }
        
        return moduleInit(identifier, namespace, preferredEnvironment, sources)
    }
}

// MARK: AssetsExplorer

private struct AssetsExplorer: DynamicSourcesExplorer {
    typealias T = JSModule.Asset
    
    static let assetsURL: URL = Bundle.main.url(forResource: "public", withExtension: nil)!.appendingPathComponent("assets")
    private static let script: String = "native/isolated_modules/isolated-modules.script.js"
    private static let modulesDir: String = "protocol_modules"
    
    private let fileManager: FileManager
    
    init(fileManager: FileManager) {
        self.fileManager = fileManager
    }
    
    func readIsolatedModulesScript() throws -> Data {
        try readData(atPath: Self.script)
    }
    
    func listModules() throws -> [String] {
        let url = Self.assetsURL.appendingPathComponent(Self.modulesDir)
        return try fileManager.contentsOfDirectory(atPath: url.path)
    }
    
    func modulePath(_ module: String, forPath path: String) throws -> String {
        "\(Self.modulesDir)/\(module)/\(path)"
    }
    
    func readModuleSources(_ module: JSModule.Asset) throws -> [Data] {
        try module.sources.lazy.map { try readData(atPath: modulePath(module.identifier, forPath: $0)) }
    }
    
    func readModuleManifest(_ module: String) throws -> Data {
        try readData(atPath: modulePath(module, forPath: FileExplorer.manifestFilename))
    }
    
    private func readData(atPath pathComponent: String) throws -> Data {
        let url = Self.assetsURL.appendingPathComponent(pathComponent)
        return try fileManager.contents(at: url)
    }
}

// MARK: DocumentExplorer

private struct DocumentExplorer: DynamicSourcesExplorer {
    typealias T = JSModule.Instsalled
    
    private static let modulesDir: String = "protocol_modules"
    
    private let fileManager: FileManager
    
    init(fileManager: FileManager) {
        self.fileManager = fileManager
    }
    
    func listModules() throws -> [String] {
        guard let url = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first else {
            return []
        }
        
        let modulesDirPath = url.appendingPathComponent(Self.modulesDir).path
        guard fileManager.fileExists(atPath: modulesDirPath) else {
            return []
        }
        
        return try fileManager.contentsOfDirectory(atPath: modulesDirPath)
    }
    
    func modulePath(_ module: String, ofPath path: String) throws -> String {
        return "\(Self.modulesDir)/\(module)/\(path)"
    }
    
    func readModuleSources(_ module: JSModule.Instsalled) throws -> [Data] {
        try module.sources.lazy.map { try readData(atPath: modulePath(module.identifier, ofPath: $0)) }
    }
    
    func readModuleManifest(_ module: String) throws -> Data {
        try readData(atPath: modulePath(module, ofPath: FileExplorer.manifestFilename))
    }
    
    private func readData(atPath path: String) throws -> Data {
        guard let url = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first?.appendingPathComponent(path) else {
            throw Error.invalidDirectory
        }
        
        return try fileManager.contents(at: url)
    }
}

// MARK: DynamicSourcesExplorer

private protocol DynamicSourcesExplorer {
    associatedtype T
    
    func listModules() throws -> [String]
    
    func readModuleSources(_ module: T) throws -> [Data]
    func readModuleManifest(_ module: String) throws -> Data
}

// MARK: Extensions

private extension FileExplorer {
    static let manifestFilename: String = "manifest.json"
}

private extension FileManager {
    func contents(at url: URL) throws -> Data {
        guard let data = contents(atPath: url.path) else {
            throw Error.invalidPath
        }
        
        return data
    }
}

private enum Error: Swift.Error {
    case invalidPath
    case invalidDirectory
}

File: android/app/src/main/java/it/airgap/vault/plugin/isolatedmodules/FileExplorer.kt
package it.airgap.vault.plugin.isolatedmodules

import android.content.Context
import android.os.Build
import com.getcapacitor.JSObject
import it.airgap.vault.plugin.isolatedmodules.js.JSModule
import it.airgap.vault.plugin.isolatedmodules.js.environment.JSEnvironment
import it.airgap.vault.util.Directory
import it.airgap.vault.util.getDirectory
import it.airgap.vault.util.readBytes
import java.io.File
import java.nio.file.Files
import java.nio.file.attribute.BasicFileAttributes

interface StaticSourcesExplorer {
    fun readJavaScriptEngineUtils(): ByteArray
    fun readIsolatedModulesScript(): ByteArray
}

interface DynamicSourcesExplorer {
    fun removeModules(identifiers: List<String>)
    fun removeAllModules()

    fun getInstalledTimestamp(identifier: String): String
}

interface SourcesExplorer<in M : JSModule> {
    fun listModules(): List<String>

    fun readModuleSources(module: M): Sequence<ByteArray>
    fun readModuleManifest(module: String): ByteArray
}

private const val MANIFEST_FILENAME = "manifest.json"

private typealias JSModuleConstructor<T> = (identifier: String, namespace: String?, preferredEnvironment: JSEnvironment.Type, paths: List<String>) -> T

class FileExplorer private constructor(
    private val context: Context,
    private val assetsExplorer: AssetsExplorer,
    private val filesExplorer: FilesExplorer,
) : StaticSourcesExplorer by assetsExplorer {
    constructor(context: Context) : this(context, AssetsExplorer(context), FilesExplorer(context))

    fun loadAssetModules(): List<JSModule.Asset> = loadModules(assetsExplorer, JSModule.Asset.constructor)

    fun loadInstalledModules(): List<JSModule.Installed> = loadModules(filesExplorer, JSModule.Installed.constructor)

    fun loadInstalledModule(identifier: String): JSModule.Installed {
        val manifest = JSObject(filesExplorer.readModuleManifest(identifier).decodeToString())

        return loadModule(identifier, manifest, JSModule.Installed.constructor)
    }

    fun loadPreviewModule(path: String, directory: Directory): JSModule.Preview {
        val moduleDir = File(context.getDirectory(directory), path)
        val manifest = JSObject(File(moduleDir, MANIFEST_FILENAME).readBytes().decodeToString())

        return loadModule(moduleDir.name, manifest, JSModule.Preview.constructor(moduleDir))
    }

    fun removeInstalledModules(identifiers: List<String>) {
        filesExplorer.removeModules(identifiers)
    }

    fun removeAllInstalledModules() {
        filesExplorer.removeAllModules()
    }

    fun readModuleSources(module: JSModule): Sequence<ByteArray> =
        when (module) {
            is JSModule.Asset -> assetsExplorer.readModuleSources(module)
            is JSModule.Installed -> filesExplorer.readModuleSources(module)
            is JSModule.Preview -> module.sources.asSequence().map { File(module.path, it).readBytes() }
        }

    fun readModuleManifest(module: JSModule): ByteArray =
        when (module) {
            is JSModule.Asset -> assetsExplorer.readModuleManifest(module.identifier)
            is JSModule.Installed -> filesExplorer.readModuleManifest(module.identifier)
            is JSModule.Preview -> File(module.path, MANIFEST_FILENAME).readBytes()
        }

    private val JSModule.Asset.Companion.constructor: JSModuleConstructor<JSModule.Asset>
        get() = JSModule::Asset

    private val JSModule.Installed.Companion.constructor: JSModuleConstructor<JSModule.Installed>
        get() = { identifier, namespace, preferredEnvironment, paths ->
            JSModule.Installed(
                identifier,
                namespace,
                preferredEnvironment,
                paths,
                filesExplorer.getInstalledTimestamp(identifier)
            )
        }

    private fun JSModule.Preview.Companion.constructor(moduleDir: File): JSModuleConstructor<JSModule.Preview> = { identifier, namespace, preferredEnvironment, paths ->
            JSModule.Preview(
                identifier,
                namespace,
                preferredEnvironment,
                paths,
                moduleDir.absolutePath
            )
        }

    private fun <T : JSModule> loadModules(
        explorer: SourcesExplorer<T>,
        constructor: JSModuleConstructor<T>,
    ): List<T> = explorer.listModules().map { module ->
        val manifest = JSObject(explorer.readModuleManifest(module).decodeToString())
        loadModule(module, manifest, constructor)
    }

    private fun <T : JSModule> loadModule(
        identifier: String,
        manifest: JSObject,
        constructor: JSModuleConstructor<T>,
    ): T {
        val namespace = manifest.getJSObject("src")?.getString("namespace")
        val preferredEnvironment = manifest.getJSObject("jsenv")?.getString("android")?.let { JSEnvironment.Type.fromString(it) } ?: JSEnvironment.Type.JavaScriptEngine
        val sources = buildList {
            val include = manifest.getJSONArray("include")
            for (i in 0 until include.length()) {
                val source = include.getString(i).takeIf { it.endsWith(".js") } ?: continue
                add(source.trimStart('/'))
            }
        }

        return constructor(identifier, namespace, preferredEnvironment, sources)
    }
}

private class AssetsExplorer(private val context: Context) : StaticSourcesExplorer, SourcesExplorer<JSModule.Asset>  {
    override fun readJavaScriptEngineUtils(): ByteArray = context.assets.readBytes(JAVA_SCRIPT_ENGINE_UTILS)
    override fun readIsolatedModulesScript(): ByteArray = context.assets.readBytes(SCRIPT)

    override fun listModules(): List<String> = context.assets.list(MODULES_DIR)?.toList() ?: emptyList()

    override fun readModuleSources(module: JSModule.Asset): Sequence<ByteArray> =
        module.sources.asSequence().map { context.assets.readBytes(modulePath(module.identifier, it))}
    override fun readModuleManifest(module: String): ByteArray = context.assets.readBytes(modulePath(module, MANIFEST_FILENAME))

    private fun modulePath(module: String, path: String): String =
        "${MODULES_DIR}/${module.trimStart('/')}/${path.trimStart('/')}"

    companion object {
        private const val SCRIPT = "public/assets/native/isolated_modules/isolated-modules.script.js"
        private const val JAVA_SCRIPT_ENGINE_UTILS = "public/assets/native/isolated_modules/isolated-modules.js-engine-android.js"

        private const val MODULES_DIR = "public/assets/protocol_modules"
    }
}

private class FilesExplorer(private val context: Context) : DynamicSourcesExplorer, SourcesExplorer<JSModule.Installed>  {
    private val modulesDir: File
        get() = File(context.filesDir, MODULES_DIR)

    override fun removeModules(identifiers: List<String>) {
        identifiers.forEach {
            File(modulesDir, it).deleteRecursively()
        }
    }

    override fun removeAllModules() {
        modulesDir.deleteRecursively()
    }

    override fun getInstalledTimestamp(identifier: String): String {
        val file = File(modulesDir, identifier)

        val timestamp = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val attributes = Files.readAttributes(file.toPath(), BasicFileAttributes::class.java)
            attributes.creationTime().toMillis()
        } else file.lastModified()

        return timestamp.toString()
    }

    override fun listModules(): List<String> = modulesDir.list()?.toList() ?: emptyList()

    override fun readModuleSources(module: JSModule.Installed): Sequence<ByteArray> =
        module.sources.asSequence().map { File(modulesDir, modulePath(module.identifier, it)).readBytes() }
    override fun readModuleManifest(module: String): ByteArray = File(modulesDir, modulePath(module, MANIFEST_FILENAME)).readBytes()

    private fun modulePath(module: String, path: String): String =
        "${module.trimStart('/')}/${path.trimStart('/')}"

    companion object {
        private const val MODULES_DIR = "protocol_modules"
    }
}